### PR TITLE
ARROW-2522: [C++] Version shared library files

### PR DESCRIPTION
**Pull request by pitrou**

### CODE REVIEW

**Comment by pitrou** [comment]
See question on JIRA about the naming scheme.

File: cpp/CMakeLists.txt
Comment (kou) [inline cpp/CMakeLists.txt]: We're using `ARROW_ABI_VERSION` in `cpp/src/arrow/arrow.pc.in`.
Please keep the name.

**Review by kou — CHANGES_REQUESTED** [review]

File: cpp/CMakeLists.txt
Comment (pitrou) [reply cpp/CMakeLists.txt]: "ABI version" is really the wrong name for this. The ABI version doesn't change every patch release.

File: cpp/CMakeLists.txt
Comment (xhochy) [inline cpp/CMakeLists.txt]: Wouldn't we be able to read this from `ARROW_VERSION`? I dislike hardcoding the version number.

**Review by xhochy — CHANGES_REQUESTED** [review]

File: cpp/CMakeLists.txt
Comment (kou) [reply cpp/CMakeLists.txt]: OK.
We have similar codes for Plasma in cpp/src/CMakeLists.txt and cpp/src/plasma/plasma.pc.in. It's better that we change them too.

Comment (kou) [reply cpp/CMakeLists.txt]: One more thing.
I did the same thing (`PARQUET_ABI_VERSION` was introduced) for parquet-cpp. It may be better that we change them too for consistency. We have a code that depends on the in `cpp/cmake_modules/FindParquet.cmake`.

Comment (pitrou) [reply cpp/CMakeLists.txt]: Note that `PLASMA_VERSION` is used in `plasma.pc.in` but never defined anywhere...

**Comment by pitrou** [comment]
I think I addressed the review comments now, could you take another look?

**Review by xhochy — APPROVED** [review]
+1, LGTM

But please wait for a review from @kou before merging

**Review by kou — APPROVED** [review]

**Comment by kou** [comment]
+1
I'll merge this!

**Pull request closed without merging**
